### PR TITLE
planner/core: match prefix columns when choosing index join

### PR DESCRIPTION
**Pull request by jarvys**
move useless equal conditions to other conditions for generating index join plan

### What problem does this PR solve? 

For #9562 

### What is changed and how it works?

If some fake condition is not useful for index join, modify the join key's idxOff to -1 and filter out the fake condition from the remained conditions. 
When constructing index join plan, the join key's equal condition will be moved to other conditions.

### Check List 

Tests 

 - Unit test 

Code changes

- None

Side effects

- None

Related changes

- None

### CODE REVIEW

File: planner/core/exhaust_physical_plans.go [hunk 1]
```diff
@@ -547,14 +547,28 @@ func (p *LogicalJoin) buildRangeForIndexJoin(indexInfo *model.IndexInfo, innerPl
 		return nil, nil, nil
 	}
 
-	// We should guarantee that all the join's equal condition is used.
-	for _, eqCond := range eqConds {
+	// Guarantee res.AccessConds is not empty.
+	if len(res.AccessConds) == 0 {
```
Comment (eurekaka) [inline planner/core/exhaust_physical_plans.go]: Is it possible that we fall into this branch?

File: planner/core/exhaust_physical_plans.go [hunk 2]
```diff
@@ -547,14 +547,28 @@ func (p *LogicalJoin) buildRangeForIndexJoin(indexInfo *model.IndexInfo, innerPl
...
 		if !expression.Contains(res.AccessConds, eqCond) {
-			return nil, nil, nil
+			keyOff2IdxOff[i] = -1
+			invalidFakeConds = append(invalidFakeConds, eqCond)
+		}
+	}
+
+	// Filter out invalidFakeConds from res.RemainedConds.
+	for _, cond := range res.RemainedConds {
+		if !expression.Contains(invalidFakeConds, cond) {
```
Comment (eurekaka) [inline planner/core/exhaust_physical_plans.go]: Why should we add this check? Why we cannot just append `res.RemainedConds` into `remained`? could you please give an example for illustration?

Comment (eurekaka) [reply planner/core/exhaust_physical_plans.go]: Oh, I see, it contains fake equal conditions, so it should be removed.

File: planner/core/exhaust_physical_plans.go (hunk 1 above)
Comment (jarvys) [reply planner/core/exhaust_physical_plans.go]: Yes, have a look at this sql:)
```sql
create table t(a int, b int, index idx_ab(a, b));
explain select * from t t1 join t t2 where t1.b = t2.b and t2.b is null
```

**Pull request merged**